Panic on NewWatcher() during reconnect

Reproducer (as a test):
```diff
diff --git a/tarantool_test.go b/tarantool_test.go
index 2be3e79..91d50aa 100644
--- a/tarantool_test.go
+++ b/tarantool_test.go
@@ -3550,6 +3550,44 @@ func TestConnection_NewWatcher(t *testing.T) {
 	}
 }
 
+func TestNewWatcherDuringReconnect(t *testing.T) {
+	const server = "127.0.0.1:3015"
+	testDialer := dialer
+	testDialer.Address = server
+
+	inst, err := test_helpers.StartTarantool(test_helpers.StartOpts{
+		Dialer:       testDialer,
+		InitScript:   "config.lua",
+		Listen:       server,
+		WaitStart:    100 * time.Millisecond,
+		ConnectRetry: 10,
+		RetryTimeout: 500 * time.Millisecond,
+	})
+	defer test_helpers.StopTarantoolWithCleanup(inst)
+	if err != nil {
+		t.Fatalf("Unable to start Tarantool: %s", err)
+	}
+
+	ctx, cancel := test_helpers.GetConnectContext()
+	defer cancel()
+
+	reconnectOpts := opts
+	reconnectOpts.Reconnect = 100 * time.Millisecond
+	reconnectOpts.MaxReconnects = 100
+	conn, err := Connect(ctx, testDialer, reconnectOpts)
+	if err != nil {
+		t.Fatalf("Connection was not established: %v", err)
+	}
+	defer conn.Close()
+	test_helpers.StopTarantool(inst)
+
+	time.Sleep(time.Second)
+	_, err = conn.NewWatcher("one", func(event WatchEvent) {})
+	if err == nil {
+		t.Fatal("Expected an error")
+	}
+}
+
 func TestConnection_NewWatcher_noWatchersFeature(t *testing.T) {
 	test_helpers.SkipIfWatchersSupported(t)
 
```

Result:
```console
mergen@megatron:~/work/go-tarantool$ go test -run TestNewWatcherDuringReconnect
2025/03/28 12:54:46 Tarantool "127.0.0.1:3013" was unexpectedly terminated: exit status 1
2025/03/28 12:54:46 tarantool: reconnect (0/100) to 127.0.0.1:3015 failed: failed to dial: dial tcp 127.0.0.1:3015: connect: connection refused
2025/03/28 12:54:46 tarantool: reconnect (1/100) to 127.0.0.1:3015 failed: failed to dial: dial tcp 127.0.0.1:3015: connect: connection refused
2025/03/28 12:54:46 tarantool: reconnect (2/100) to 127.0.0.1:3015 failed: failed to dial: dial tcp 127.0.0.1:3015: connect: connection refused
2025/03/28 12:54:47 tarantool: reconnect (3/100) to 127.0.0.1:3015 failed: failed to dial: dial tcp 127.0.0.1:3015: connect: connection refused
2025/03/28 12:54:47 tarantool: reconnect (4/100) to 127.0.0.1:3015 failed: failed to dial: dial tcp 127.0.0.1:3015: connect: connection refused
2025/03/28 12:54:47 tarantool: reconnect (5/100) to 127.0.0.1:3015 failed: failed to dial: dial tcp 127.0.0.1:3015: connect: connection refused
2025/03/28 12:54:47 tarantool: reconnect (6/100) to 127.0.0.1:3015 failed: failed to dial: dial tcp 127.0.0.1:3015: connect: connection refused
2025/03/28 12:54:47 tarantool: reconnect (7/100) to 127.0.0.1:3015 failed: failed to dial: dial tcp 127.0.0.1:3015: connect: connection refused
2025/03/28 12:54:47 tarantool: reconnect (8/100) to 127.0.0.1:3015 failed: failed to dial: dial tcp 127.0.0.1:3015: connect: connection refused
2025/03/28 12:54:47 tarantool: reconnect (9/100) to 127.0.0.1:3015 failed: failed to dial: dial tcp 127.0.0.1:3015: connect: connection refused
--- FAIL: TestNewWatcherDuringReconnect (1.62s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x5922e2]

goroutine 35 [running]:
testing.tRunner.func1.2({0x67f920, 0x8f05e0})
	/usr/lib/go-1.22/src/testing/testing.go:1631 +0x24a
testing.tRunner.func1()
	/usr/lib/go-1.22/src/testing/testing.go:1634 +0x377
panic({0x67f920?, 0x8f05e0?})
	/usr/lib/go-1.22/src/runtime/panic.go:770 +0x132
github.com/tarantool/go-tarantool/v2.(*Connection).NewWatcher(0xc00019e180, {0x6c35f7, 0x3}, 0x6f7350)
	/home/mergen/work/go-tarantool/connection.go:1422 +0x42
github.com/tarantool/go-tarantool/v2_test.TestNewWatcherDuringReconnect(0xc0003401a0)
	/home/mergen/work/go-tarantool/tarantool_test.go:3585 +0x365
testing.tRunner(0xc0003401a0, 0x6f6f60)
	/usr/lib/go-1.22/src/testing/testing.go:1689 +0xfb
created by testing.(*T).Run in goroutine 1
	/usr/lib/go-1.22/src/testing/testing.go:1742 +0x390
exit status 2
FAIL	github.com/tarantool/go-tarantool/v2	1.735s
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -14,6 +14,9 @@ Versioning](http://semver.org/spec/v2.0.0.html) except to the first release.
 
 ### Fixed
 
+- Fixed panic when calling NewWatcher() during reconnection or after
+  connection is closed (#438).
+
 ## [v2.3.2] - 2025-04-14
 
 This release improves the logic of `Connect` and `pool.Connect` in case of a

--- a/connection.go
+++ b/connection.go
@@ -1461,7 +1461,7 @@ func (conn *Connection) NewWatcher(key string, callback WatchCallback) (Watcher,
 	// That's why we can't just check the Tarantool response for an unsupported
 	// request error.
 	if !isFeatureInSlice(iproto.IPROTO_FEATURE_WATCHERS,
-		conn.c.ProtocolInfo().Features) {
+		conn.serverProtocolInfo.Features) {
 		err := fmt.Errorf("the feature %s must be supported by connection "+
 			"to create a watcher", iproto.IPROTO_FEATURE_WATCHERS)
 		return nil, err

--- a/tarantool_test.go
+++ b/tarantool_test.go
@@ -3550,6 +3550,72 @@ func TestConnection_NewWatcher(t *testing.T) {
 	}
 }
 
+func newWatcherReconnectionPrepareTestConnection(t *testing.T) (*Connection, context.CancelFunc) {
+	t.Helper()
+
+	const server = "127.0.0.1:3015"
+	testDialer := dialer
+	testDialer.Address = server
+
+	inst, err := test_helpers.StartTarantool(test_helpers.StartOpts{
+		Dialer:       testDialer,
+		InitScript:   "config.lua",
+		Listen:       server,
+		WaitStart:    100 * time.Millisecond,
+		ConnectRetry: 10,
+		RetryTimeout: 500 * time.Millisecond,
+	})
+	t.Cleanup(func() { test_helpers.StopTarantoolWithCleanup(inst) })
+	if err != nil {
+		t.Fatalf("Unable to start Tarantool: %s", err)
+	}
+
+	ctx, cancel := test_helpers.GetConnectContext()
+
+	reconnectOpts := opts
+	reconnectOpts.Reconnect = 100 * time.Millisecond
+	reconnectOpts.MaxReconnects = 0
+	reconnectOpts.Notify = make(chan ConnEvent)
+	conn, err := Connect(ctx, testDialer, reconnectOpts)
+	if err != nil {
+		t.Fatalf("Connection was not established: %v", err)
+	}
+
+	test_helpers.StopTarantool(inst)
+
+	// Wait for reconnection process to be started.
+	for conn.ConnectedNow() {
+		time.Sleep(100 * time.Millisecond)
+	}
+
+	return conn, cancel
+}
+
+func TestNewWatcherDuringReconnect(t *testing.T) {
+	test_helpers.SkipIfWatchersUnsupported(t)
+
+	conn, cancel := newWatcherReconnectionPrepareTestConnection(t)
+	defer conn.Close()
+	defer cancel()
+
+	_, err := conn.NewWatcher("one", func(event WatchEvent) {})
+	assert.NotNil(t, err)
+	assert.ErrorContains(t, err, "client connection is not ready")
+}
+
+func TestNewWatcherAfterClose(t *testing.T) {
+	test_helpers.SkipIfWatchersUnsupported(t)
+
+	conn, cancel := newWatcherReconnectionPrepareTestConnection(t)
+	defer cancel()
+
+	_ = conn.Close()
+
+	_, err := conn.NewWatcher("one", func(event WatchEvent) {})
+	assert.NotNil(t, err)
+	assert.ErrorContains(t, err, "using closed connection")
+}
+
 func TestConnection_NewWatcher_noWatchersFeature(t *testing.T) {
 	test_helpers.SkipIfWatchersSupported(t)
 
@@ -4149,12 +4215,12 @@ func runTestMain(m *testing.M) int {
 	startOpts.MemtxUseMvccEngine = !isStreamUnsupported
 
 	inst, err := test_helpers.StartTarantool(startOpts)
-	defer test_helpers.StopTarantoolWithCleanup(inst)
-
 	if err != nil {
 		log.Printf("Failed to prepare test tarantool: %s", err)
 		return 1
 	}
+
+	defer test_helpers.StopTarantoolWithCleanup(inst)
 
 	return m.Run()
 }

